doc/decisions/array.md

```diff
@@ -3,26 +3,25 @@
 ## Problem
 
 Currently it is inefficient to detect the length of an array and
-it is impossible to know if an key (without subkeys) should be
+it is impossible to know if a key (without subkeys) should be
 an array or not.
 
-For the latter problem different workarounds exist, such as
-`###empty_array` in `yajl`.
-
 ## Constraints
 
 - None
 
 ## Assumptions
 
-- None
+- Meta-data arrays simply work by convention as they are not serialized in special ways nor they get validated.
+  The convention will be the same but without any metadata marker.
 
 ## Considered Alternatives
 
 - `###empty_array` as in `yajl`, problem: does not allow efficient access of first element
 - store length (and not last element), problem: needs prepending of `#_...`
-- store element after last element (C++-Style), problem: very unusual style
-- use value and not metadata array, problem: is ambiguous
+- store element after last element (C++-Style), would not fit nicely with key/value
+- use value and not the metadata `array`, problem: is ambiguous
+- use metadata on all children
```

Comment:
Continuation from https://github.com/ElektraInitiative/libelektra/issues/3223#issuecomment-712661004
---

> Yes, spec would really need to understand the semantics of the metadata, like arrays. So it would check all namespaces how the array actually looks like, and then add the proper metadata array.

I still don't understand how that would work. Without any metadata in `user:/` there can be no array elements in `user:/` according to the current definition of "array element".

For example, the `spec` plugin would find `spec:/myarray` with an `array` metadata. It would therefore have to validate this array. It then does a lookup on `/myarray` to find the actual array size. But that will never return any `array` metadata, if the storage format doesn't support metadata. Therefore `spec` always assumes `/myarray` should be of the default size defined in `spec:/myarray`.

Reply:
The other way round: it would add the size 692ce07d4be4be3fd4961aedaa65862f25c2f91a

Reply:
Relying on the `spec` for all of this will also cause some problems with plugin ordering.

We need to run `spec` first to add the correct size, but we need to run it last to validate against broken arrays. Actually, we also need to run it between all other plugins for the validation. Otherwise plugins cannot make the assumption that there are no broken arrays.